src/aws-cpp-sdk-core/include/aws/core/utils/crypto/CRC.h

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+
+#include <aws/core/utils/crypto/Hash.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/logging/LogMacros.h>
+#include <type_traits>
+
+namespace Aws
+{
+    namespace Utils
+    {
+        namespace Crypto
+        {
+            template <typename HashT>
+            ByteBuffer ConvertToBuffer(HashT value)
+            {
+                static_assert(std::is_integral<HashT>::value, "Must use integral type to convert to buffer");
+                Aws::Utils::ByteBuffer buffer(sizeof(HashT));
+                for (size_t i = 0; i < sizeof(HashT); ++i)
+                {
+                    size_t shiftSize = (sizeof(HashT) - 1) * 8 - i * 8;
+                    buffer[i] = (value >> shiftSize) & 0xFF;
+                }
+                return buffer;
+            }
+
+            template <typename RunningChecksumT,
+                      RunningChecksumT (*CRTChecksumFuncT)(Crt::ByteCursor, RunningChecksumT),
+                      ByteBuffer (*ByteBufferFuncT)(RunningChecksumT)>
+            class CRCChecksum : public Hash
+            {
+            public:
+                CRCChecksum() : m_runningChecksum{0}
+                {
+                }
+
+                ~CRCChecksum() override = default;
+
+                HashResult Calculate(const Aws::String& str) override
+                {
+                    Aws::Crt::ByteCursor byteCursor = Aws::Crt::ByteCursorFromArray(reinterpret_cast<const uint8_t*>(str.data()), str.size());
+                    m_runningChecksum = CRTChecksumFuncT(byteCursor, m_runningChecksum);
+                    return ByteBufferFuncT(m_runningChecksum);
+                };
+
+                HashResult Calculate(Aws::IStream& stream) override
+                {
+                    auto currentPos = stream.tellg();
+                    if (currentPos == std::ios::pos_type(-1))
+                    {
+                        currentPos = 0;
+                        stream.clear();
+                    }
+
+                    stream.seekg(0, Aws::IStream::beg);
+
+                    uint8_t streamBuffer[Aws::Utils::Crypto::Hash::INTERNAL_HASH_STREAM_BUFFER_SIZE];
+                    while (stream.good())
+                    {
+                        stream.read(reinterpret_cast<char*>(streamBuffer), Aws::Utils::Crypto::Hash::INTERNAL_HASH_STREAM_BUFFER_SIZE);
```

Comment:
better to use 
while (stream.read(reinterpret_cast<char*>(streamBuffer), Aws::Utils::Crypto::Hash::INTERNAL_HASH_STREAM_BUFFER_SIZE)) {
to handle end of stream errors more cleanly

Reply:
this actually causes the integration tests to fail, so there is a behavior difference between `stream.good` and `if(stream.read)`